python.d/unbound.chart.py

```diff
@@ -0,0 +1,165 @@
+        self.ext = self.configuration.get('extended', None)
+        self.ubconf = self.configuration.get('ubconf', None)
+        self.order = ORDER
+        self.definitions = CHARTS
+        if self.ext:
+            self.order = self.order + EXTENDED_ORDER
+            self.definitions.update(EXTENDED_CHARTS)
+        self.request = 'UBCT1 stats\n'
+        self._parse_config()
+        self.debug('Unbound config: {0}'.format(self.ubconf))
+        self._auto_config()
+        self.debug('Extended stats: {0}'.format(self.ext))
+        if self.unix_socket:
+            self.debug('Using unix socket: {0}'.format(self.unix_socket))
+            for key in self.definitions.keys():
+                self.definitions[key]['options'][4] = 'Local'
+        else:
+            self.debug('Connecting to: {0}:{1}'.format(self.host, self.port))
+            self.debug('Using key: {0}'.format(self.key))
+            self.debug('Using certificate: {0}'.format(self.cert))
+            for key in self.definitions.keys():
+                self.definitions[key]['options'][4] = self.host
+
+    def _auto_config(self):
+        if self.ubconf and os.access(self.ubconf, os.R_OK):
+            conf = YamlOrderedLoader.load_config_from_file(self.ubconf)[0]
+            if self.ext is None:
+                if 'extended-statistics' in conf['server'].keys():
+                    self.ext = conf['server']['extended-statistics']
+            if 'remote-control' in conf.keys():
+                if conf['remote-control'].get('control-use-cert', False):
+                    if not self.key:
+                        self.key = conf['remote-control'].get('control-key-file')
+                    if not self.cert:
+                        self.cert = conf['remote-control'].get('control-cert-file')
+                    if not self.port:
+                        self.port = conf['remote-control'].get('control-port')
+                else:
+                    if not self.unix_socket:
+                        self.unix_socket = conf['remote-control'].get('control-interface')
+        if not self.key:
+            self.key = '/etc/unbound/unbound_control.key'
+        if not self.cert:
+            self.cert = '/etc/unbound/unbound_control.pem'
+        if not self.port:
+            self.port = 8953
+
+    def check(self):
+        # We need to check that auth works, otherwise there's no point.
+        self._connect()
+        result = bool(self._sock)
+        self._disconnect()
+        return result
+
+    def _check_raw_data(self, data):
+        # The server will close the connection when it's done sending
+        # data, so just keep looping until that happens.
+        return False
+
+    def _get_data(self):
+        raw = self._get_raw_data()
+        data = dict()
+        tmp = dict()
+        for line in raw.splitlines():
+            stat = line.split('=')
+            tmp[stat[0]] = stat[1]
+        for item in STAT_MAP.keys():
+            if item in tmp.keys():
```

Comment:
`.keys()` is not needed, just do `for item in DICT` (same for `if item in DICT`)
this is not the only place you use `dict.keys()`

Reply:
This is mostly a habit from where I work (most of our programmers don't do much Python coding, and stuff like this helps a bit clarity-wise for some of them).  I'll push a commit to convert all of these shortly.

Reply:
I am not a programmer but that is a bad habbit. `dict.keys()` produces a new list (py2) and `in` list is much slower than `in` dict.